LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings

acrowthe opened the pull request:
Implement new vocabularyList questionnaire editor type, enabling a multi-select dropdown of all installed vocabularies for vocabulary questions.

Note: support for vocabulary filters in the wizard is not yet covered by this PR.

acrowthe committed:
LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings
Initial implentation: Copy ListInput for ease of implementation.
- Future commits or PRs will refactor list input

acrowthe committed:
LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings
Fix issues with vocabularyList and questionnaire editor property definition

marta- reviewed (commented):
Works great. Refactoring to merge with ListInput is the next step.

acrowthe committed:
LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings
Refactor logic from VocabularyInput into ListInput
- This input is used when an input description starts with "list"
- SubjectTypes or Vocabulary can be specified by "list.vocabularies" or "list.subjectTypes"
- Defaults to subjectTypes to preserve original functionality of "list"

acrowthe commented:
ListInput and VocabularyInput have been merged together, and this PR is now ready for review and testing

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ListInput.jsx:
```diff
@@ -25,22 +25,52 @@ import EditorInput from "./EditorInput";
+  static displayVariable = "label";
+  static orderVariable = "lfs:defaultOrder";
+  static saveVariable = "jcr:uuid";
+  static uniqueIdentifier = "jcr:uuid";
```
I think it may be safe to assume that `jcs:uuid` is always thr unique identifier

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/VocabularyInput.jsx:
```diff
@@ -0,0 +1,108 @@
+import QuestionnaireStyle from '../questionnaire/QuestionnaireStyle';
+import QuestionComponentManager from "./QuestionComponentManager";
+
+let VocabularyInput = (props) => {
```
This component can be removed now

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ListInput.jsx:
```diff
@@ -25,22 +25,52 @@ import EditorInput from "./EditorInput";
 import QuestionnaireStyle from '../questionnaire/QuestionnaireStyle';
 import QuestionComponentManager from "./QuestionComponentManager";
 
+class VocabularyDefinition {
+  static primaryType = "lfs:Vocabulary";
+  static saveType = "String";
+  static displayVariable = "identifier";
+  static orderVariable = this.primaryType;
+  static saveVariable = this.displayVariable;
+  static uniqueIdentifier = "jcr:uuid";
+}
+class SubjectTypeDefinition {
+  static primaryType = "lfs:SubjectType";
+  static saveType = "Reference";
+  static displayVariable = "label";
+  static orderVariable = "lfs:defaultOrder";
+  static saveVariable = "jcr:uuid";
+  static uniqueIdentifier = "jcr:uuid";
+}
```
I wonder if we can make this more extensible by moving the details of the query to the configuration json

acrowthe replied:
Moved into Properties.json and Question.json:
```json
"sourceVocabularies": {
          "type": "list",
          "primaryType": "lfs:Vocabulary",
          "saveType": "String",
          "displayVariable": "identifier",
          "orderVariable": "lfs:Vocabulary",
          "saveVariable": "identifier"
        },
 ```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ListInput.jsx:
```diff
@@ -25,22 +25,52 @@ import EditorInput from "./EditorInput";
+  let { objectKey, data, value: fieldDefinition } = props;
   let [ value, setValue ] = React.useState(Array.isArray(data[objectKey]) ? data[objectKey] : data[objectKey] ? [data[objectKey]] : []);
   const [ options, setOptions ] = React.useState([]);
   const requiredSubjectTypes = React.useState(objectKey.includes('requiredSubjectTypes'));
```
This looks specific to `requiredSubjectTypes`, shouldn't it be either generalized or removed?

acrowthe replied:
Removed

sashaandjic commented:
Problem found
Steps to reproduce:
- install several vocabularies
- modify Family field of Demographic form to use 2 vocabularies
![image](https://user-images.githubusercontent.com/23243244/106658034-c5899900-656a-11eb-98de-0e4527e15033.png)
- save it and observe that both vocabularies are shown
![image](https://user-images.githubusercontent.com/23243244/106658096-d89c6900-656a-11eb-8464-c24762da20ad.png)
- click on the edit (pencil) button
- notice that no vocabularies are listed
![image](https://user-images.githubusercontent.com/23243244/106658238-071a4400-656b-11eb-8e20-c4371c251852.png)
- if you click on SAVE you will lose previously selected vocabularies

acrowthe committed:
LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings
Further clean up listInput
- Move query variables into json configuration files
- Remove temporary VocabularyInput
- Clean up extraneous subjectType code

acrowthe committed:
LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings
Remove testing log message

sashaandjic reviewed (approved):
tested - all good

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ListInput.jsx:
```diff
@@ -101,7 +101,8 @@ var StyledListInput = withStyles(QuestionnaireStyle)(ListInput);
 QuestionComponentManager.registerQuestionComponent((definition) => {
-  if (["list"].includes(definition)) {
+  if (definition.type && definition.type === "list") {
+    // typeof(definition) === "string" && definition.startsWith("list")) {
```
Remove dead code.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Properties.json:
```diff
@@ -2,6 +2,13 @@
+      "saveType": "Reference",
+      "displayVariable": "label",
+      "orderVariable": "lfs:defaultOrder",
+      "saveVariable": "jcr:uuid"
```
These aren't exactly variables, more like properties. And I'm not fond of `save` either, I would use `displayProperty`, `orderProperty`, and `idProperty` or `identifierProperty`.

acrowthe committed:
LFS-712: Source vocabulary fields in the questionnaire editor should be arrays, not strings
Clean up commented code and rename configurable properties

sashaandjic reviewed (approved):
RE-TESTED
